torch/_inductor/fx_passes/quantization.py

```diff
@@ -588,6 +588,149 @@ def _generate_qconv_weight_prepack_patterns():
+    to_fp32_node = sub_node.args[0]
+
+    assert to_fp32_node.target is prims.convert_element_type.default
+    assert sub_node.target is aten.sub.Tensor
+    assert mul_node.target is aten.mul.Tensor
+    if (
+        len(list(to_fp32_node.users)) != 1
+        or len(list(sub_node.users)) != 1
+        or len(list(mul_node.users)) != 1
+    ):
+        # Ensure the dequant pattern only has 1 user
+        # since we will delete the dequant pattern here
+        return False
+    return True
+
+
+def _register_qlinear_weight_prepack_pass(pattern, pass_number):
+    @register_freezing_graph_pattern(
+        pattern,
+        extra_check=_is_valid_dequant_linear_pattern,
+        pass_number=pass_number,
+    )
+    def qlinear_weight_prepack(match: Match, *args, **kwargs):
+        """
+        Match the pattern:
+        int8 activation
+          |
+        dequant_per_tensor
+          |
+        mm/addmm <- t <- dequant_per_channel <- int8_weight
+
+        Insert weight prepack node and change the pattern to:
+        int8 activation
+          |
+        onednn.qlinear_pointwise <- onednn.qlinear_prepack <- int8_weight
+        """
+        linear_node = match.output_node()
+        assert linear_node.target in (aten.addmm.default, aten.mm.default)
+        input_index = 0 if linear_node.target is aten.mm.default else 1
+        weight_index = input_index + 1
+        mul_node = linear_node.args[input_index]
+        sub_node = mul_node.args[0]
+        to_fp32_node = sub_node.args[0]
+        t_node = linear_node.args[weight_index]
+        dequant_per_channel = t_node.args[0]
+        assert (
+            dequant_per_channel.target
+            is quantized_decomposed.dequantize_per_channel.default
+        )
+
+        # Activation QParams
+        qx, x_zp, x_scale = (
+            kwargs["x"],
+            kwargs["x_zp"],
+            kwargs["x_scale"],
+        )
+
+        # Weight QParams
+        qw, w_scale, w_zp = (
+            kwargs["q_weight"],
+            kwargs["w_scale"],
+            kwargs["w_zp"],
+        )
+
+        # Params
+        bias = kwargs["b"] if "b" in kwargs else None
+
+        x_shape = qx.meta.get("tensor_meta").shape
+        graph = match.graph
+        with graph.inserting_before(linear_node):
+            # Insert weight prepack node and the qlinear node
+            packed_weight_inputs = (
+                qw,
+                x_shape,
+            )
+            packed_weight_op = torch.ops.onednn.qlinear_prepack
+            prepack_weight_node = graph.call_function(
+                packed_weight_op, args=packed_weight_inputs
+            )
+
+            new_args = (
+                qx,
+                x_scale,
+                x_zp,
+                prepack_weight_node,
+                w_scale,
+                w_zp,
+                bias,
+                1.0,  # output_scale
+                0,  # output_zero_point
+                True,  # fp32_output
```

Comment:
we always output fp32 for qlinear?

Reply:
Yes. This is for weight prepack, not the final graph. The original pattern outputs fp32, so we need to keep this semantics.